### DISCUSSION TITLE
Should MapLibre have a Style Editor?

### DISCUSSION THREAD

**Discussion by wipfli** [maplibre/maplibre#153] [General]
People edit MapLibre style json documents in many different ways. OpenStreetMap Americana generates it from JavaScript (@ZeLonewolf), some use a visual editor like [Maputnik](https://github.com/maputnik/editor) (@nyurik and @lseelenbinder maybe), some create it with [Charites](https://github.com/unvt/charites) (@pnorman maybe), some use [Fresco](https://github.com/go-spatial/fresco) (@roblabs maybe) and I myself just edit the `style.json` document by hand.

The question came up if MapLibre should maintain a style editor like for example Maputnik. What do you think?

**Comment by smellyshovel** [comment]
I personally feel like there's no such a necessity. It's better to maintain what's already there (other open-source solutions like Maputnik).

On the other hand that's only true as long as the MapLibre style spec is backwards-compatible with the MapBox style spec...

**Reply by wipfli** [reply]
Thanks for the feedback.

I think it is our goal and promise to maintain compatibility of the style spec with older versions of MapLibre GL JS. So we will not introduce breaking changes in the style spec at any point of time I would say. 

What we are not able to guarantee however is compatibility with new features that Mapbox GL JS introduces after they switch to the non-permissive license in v2.

**Reply by ghost** [reply]
I think that has to go both directions for existing editors to remain useful:

- Keep mapbox feature-set of style spec working that an editor is using.
- Not introduce any new critical features in MapLibre OR (preferrably) ensure that style editors adopt our changes, so people can use the latest MapLibre features.

**Reply by lseelenbinder** [reply]
A key reason why this is currently under discussion is it appears (and I hope I'm wrong) that Maputnik is _de facto_ unmaintained today. This means a key option for designing MapLibre styles will eventually stop working entirely (it already does not support newer functionality such as 3D terrain).

**Reply by ghost** [reply]
I got the impression that starting a new tool was the idea in the first post, because it listed different workflows (which could potentially be unified in a new tool).

Here are my interpretations of the question:

- "Should we ensure that at least 1 actively maintained style editor for MapLibre exists [that non-developers / artists can use]?"
    - Definitely yes. And we should spend time (and/or money) on it if necessary. I don't care if it's maintained by MapLibre or someone else, as long as it remains FOSS. Which, I believe is also what @smellyshovel and @VA7DBI meant / understood?
- "Should MapLibre create a style-editor from scratch that's exclusively for developers?"
    - No. Because they have enough tools to generate style-specs. Artists need these tools though.
- "Should MapLibre create a style-editor from scratch that artists/developers can use?"
    - No, unless it's easy to create/maintain or huge benefits. I brought up the idea of VS Code extension below OR live benchmarking/debugging/editing of styles through a MapLibre debugger. With a good interface, both of these would also be usable by artists.

**Comment by VA7DBI** [comment]
I don't think we need to maintain our own (at last not at this time), but perhaps we could endorse one, even if it is an "if you would like to use a graphical editor, many people affiliated with the project have good things to say about ...."

**Reply by lseelenbinder** [reply]
I would much prefer to simply endorse a good graphical editor, but as I am aware, there is no _maintained_ editor that supports MapLibre, so we can't do this today.

**Comment by vinayakkulkarni** [comment]
It's basically an map authoring tool which can / will be useful for people who want to build their own map styles from scratch. it certainly would be a competitor to Mapbox Studio (eventually if the efforts are directed properly) but for doing that, we would require tons of stuff to be setup for making it purely multi-tenant app, e.g. unique mapId for each signed up user with their own data & own styles.

If we build it in public & make it FOSS, this would be a great addition but it has it's own limitations (especially in terms of cloud resources & architecting it so that it's scalable) but I personally think the pros outweigh the cons

**Comment by ghost** [comment]
I assume the board should nudge us in a direction where the project should be heading; if they think it's more of an general map-related organization (as implied by tileserver) than just a piece of software for map rendering, then it might make sense to expand.

*Personally* I think maintaining a style editor is not necessary.

Here's my take on your examples:

- OpenStreetMap Americana does also generate SVGs on the fly; they probably *must* dynamically generate data. I also hope to see their work on shields in a maplibre plugin some day, so it works with any style (~stylable shields basically, either through another style-spec, or using an API).
- Fresco/Maputnik exist and are FOSS. We should support them / add support for new MapLibre features. Multiple visual editors like Fresco and Maputnik exist and I assume at least one will remain MapLibre compatible as long as MapLibre does reasonable things.
- Charites appears to be very lightweight and mostly about being able to decompose into manual files; it doesn't appear to do any heavy lifting. It also explicitly mentions MapLibre already, so I assume it will be updated as MapLibre grows. Possibly, these tools could consume the MapBox style-spec definition to generate much of their code/data-structures, so these will still function.
- Manual editing is always available / simple, and there can't be any bitrot. It's not very visual, but gets the job done for smaller changes.

So there are different areas where you need some tool/workflow over another.
If you want more advanced features to bring different workflows together, that can probably be done by extending existing tools; but you still won't be able to replace all workflows (example: OSM Americana).

I just think it's very important that we don't break the style-spec definition (= keep the structure of it as-is, but adding or even removing features is fine). We also need good docs and examples, so developers of tools always know how a feature is supposed to work / how it could be integrated in their editor.
Keeping the changelog up-to-date is also critical if we outsource this work.

---

For completeness, here's the typical workflow at our company (traffic management/logistics planning with mapbox or maplibre being a key part in every product). I can only report on the web-team (maplibre-gl-js / mapbox-gl-js 1.x), although I believe it's similar for the mobile-team (maplibre-gl-native)

#### For the basemap:

- Artist creates a mapstyle in maputnik.

This can be a lengthy process.
Also, because our tile server must generate the spritesheets etc. this involves many people and tools aside from maputnik.

#### For application specific layers (such as realtime traffic data, known traffic anomalies, construction sites, ...):

- Developers implement vector-tile-servers for custom data.
- Artists create some layer styles (partially maputnik), which refer to these new tile-servers, and store the layer defintions on the backend.
- Frontend fetches list of these additional layers from backend and pulls the layer defintions for user-enabled layers.

This has the benefit of working on mobile + app. We can also add new content without pushing an app update.

#### For highly dynamic data which must be on the client machine anyway (such as realtime-delivery routes of a logistics company):

- Developers write code for our frontend which dynamically fills GeoJSON layers from data that's already in our model.
- Developers add hand-written styles for these is in the application code.
- Artists then often mockups are done in tools like Illustrator/After Effects/Zeplin to finalize the look.
- Developers implement these finalizing changes.

Workflow allows for fast iteration.

#### For animated layers (such as animated symbols for deliveries/pickups or the delivery vehicles):

*We used to have a mapbox fork; then a custom-layer using WebGL. But due to performance issues of re-rendering the basemap we first split our custom-layer into a separate canvas with separate WebGL context (decoupling basemap rendering and application content).
Due to high overhead of maintaining and writing our own renderer we eventually switched to deck.gl instead. So now we have mapbox / maplibre + deck.gl*

- Artists create mockup first.
- deck.gl custom layers / shaders are created as needed by me (developer), based on mockups by artists, but often adding or removing easy/hard to implement features, usually while working with the artist.
- deck.gl layer definitions are always written by hand, also because they feed our custom layers.

This is very time consuming, so we try to finalize the design early. This usually replaces one of the GeoJSON layers.

#### Summary:

So, as you can see, there are multiple people involved (backend-devs, frontend-devs, artists) and each has their own workflow.

**Comment by ZeLonewolf** [comment]
From my perspective (OSM Americana development), a static style editor that can spit out a style.json is not particularly useful to me.  I generally develop in VS Code with live reload so I can see my changes update on save.  I'm particularly challenged in my transportation layer, in which I have to generate multiple copies of bridge and tunnel layers to achieve a stacked/layering effect when bridges overlap.  We have 333 layers presently as a result of this!

However, if a style editor were able to load and edit a live map (and be able to display the JSON behind a particular layer), that would be much more useful.

Yet, if we have a choice of where to apply resources, I would much rather direct them towards the core libraries themselves to add features and improve performance.

**Reply by ghost** [reply]
I think you bring up a good point here: IDEs.
I usually preview GLTF and Wavefront OBJ files in VS Code.

I'd very much welcome an VS Code extension which auto-detects that you are editing a maplibre style (or a subset of it, for example, as part of a constant JS object).
I had a quick look if something like this exists already and there seems to be something for mapbox https://marketplace.visualstudio.com/items?itemName=AlexanderBelokon.mapbox-preview

Because there probably won't be any bitrot (unless the maplibre API changes), this feels like it's easy to implement and maintain as part of the MapLibre organization.

---

Another thing that's slightly related (non-style editors as style-editor) is a maplibre specific debugger. I have a debugger bookmarklet which searches for mapbox JS objects and instruments. This allows me to replace layers on the fly, to export their styles to the console, measure rendering time, event frequency, recording/playing camera movement, detect redundant API calls etc.
It has proven very useful and could - in theory - be used similar to a style editor in certain situations (with the benfit of also benchmarking style changes).
However, this needs very much internal knowledge about maplibre-gl-js/maplibre-gl-native, so it might not be worth doing it, unless we instrument the code with custom debug events/calls so the frontend can be simplified.

**Reply by HarelM** [reply]
@JannikGM let's try and keep this discussion focused on its title. Feel free to open different discussions on topic that are not related to "style editor". Thanks for the valuable input! :-)

**Reply by ghost** [reply]
I think the discussions are the same.
It comes down to the users workflow and intend of editing a style + wether the community is able to create an maintain such a solution.

I think VS Code *could* be a "style-editor" (that's very lightweight to maintain).
Likewise I think a debugger can also do the task of a style-editor (that's heavy to maintain, but has *much* greater benefits than an external solution as it would - for example - allow style-changes at runtime).

While I'm against "MapLibre-maintained-Maputnik/Mapbox Studio" (hard to get right + few benefits over just contributing to Maputnik) I like the idea of the organization / community maintaining a VS code extension or a MapLibre specific debugger.

**Comment by lseelenbinder** [comment]
I laid out my reasoning for why MapLibre + Maputnik is a great idea here: https://github.com/maputnik/editor/issues/788.

The salient bits are:

1) Mapbox GL JS is more naturally used in coordination with Mapbox Studio.
2) MapLibre needs a good graphical editor (and Maputnik is a good foundation to have a one).
3) The MapLibre community has resources that it seems the Maputnik community does not have, and we have a reason for it to exist.

In my mind, a rendering library without an easy, DIY way to edit styles (without opening a text editor) is close to a necessity for quick iteration by visual designers and cartographers (who prefer to avoid JSON).

**Comment by zstadler** [comment]
With the _de-facto_ freeze in extending the MapLibre style spec, it looks like there isn't much value in having a MapLibre style editor. Even the _de-facto_ lack of maintenance for Maputnik seems OK given that the spec is frozen.

I do agree with the voices here and on Slack that extending the style spec would be of great value. A repeated theme is the need for some sort of "repeat and modify" mechanism for handling multiple levels of bridges. This should, and probably could, be done in a backward-compatible way.

Once a decision is made to extend the style spec, many of us will need a style editor that supports it. I would love to see Maputnik join MapLibre and extended to support a "repeat and modify" mechanism.

**Reply by wipfli** [reply]
I think these are the main additions that the style spec has seen so far:

* https://github.com/maplibre/maplibre-gl-js/pull/347
* https://github.com/maplibre/maplibre-gl-js/pull/716
* https://github.com/maplibre/maplibre-gl-js/pull/1022
* https://github.com/maplibre/maplibre-gl-js/pull/1289

**Reply by zstadler** [reply]
Given that some extensions have been made, and the nature of these changes, I'd like to rephrase the above:

> Once a decision is made to significantly extend the style spec, such as implementing a "repeat and modify" syntax within the style, many of us will need a style editor that supports it.

**Comment by mizmay** [comment]
Hey hey, Finally coming around to offer a few cents to this conversation, because it's a really interesting one. 

First: yep totally agree with the general consensus, creating a dedicated style editor for MapLibre would be a huge time suck and a big distraction, not worth it, for reasons already mentioned.

The reason I'll add is that from a certain standpoint, style editors don't make sense, and in fact in my experience just compound the performance drags style JSONs create when they are created programmatically based on toggles and sliders instead of being optimized for render performance. This is a problem with Maputnik as well. The JSON these editors generate contain an unnecessary amount of JSON verbiage, pretty much be design, because the purpose of the editor is to produce a stylesheet quickly, not produce a stylesheet that is performant, robust, and easy to maintain.

Fancy stylesheet WYSWYGs also obfuscate the REAL performance improvement opportunities, which come from removing data you are not using from the tiles, and/or restructuring the tile metadata (style attributes) instead of using tons of stylesheet filters.

This is not to say style editors are bad -- Maputnik is awesome, and necessary for some use cases -- it's just that for a project like MapLibre, where typically you own the full stack, there's an opportunity to do things in a more transparent and flexible way. 

The workflow we've developed at Stamen is simple:
1. Visually iterate on stylesheet changes with your favorite code editor (VS Code, Sublime) and a viewer that allows for side-by-side viewing (ours is called [Maperture](https://github.com/stamen/maperture))
2. Use git and code reviews to track stylesheet changes. 

I've worked with lots of cartographers and designers who were not initially comfortable working with text editors and git, but once they have a setup and some support, and know they can fall back on loading the same stylesheet they are working with into Maputnik, quickly adapt. It's necessary and more efficient when you are working on the bleeding edge, with tile, renderer, or front-end engineers.

In the grander scheme of things, this enables cross-disciplinary cartography that isn't siloed between designers and engineers. You need that if you are developing on the bleeding edge, as is the case with every MapLibre project I've ever seen. 

Here's a non-exhaustive list of what else is we can integrate and support as a result of focusing on workflows, rather than GUIs:
1. Stylesheet linting
2. Stylesheet versioning
3. Build systems for maintain slightly varying stylesheets
5. Integrations with Figma and/or custom design tools
6. Rapid parallel iteration on other key map assets - sprites, fonts, and TILES

The last thing I'll point out, though it's probably obvious, is that ours is the workflow you need to manage the still murky intentional / unintentional divergences between Mapbox GL and MapLibre GL, and the equally murky intentional / unintentional divergences between style specs.

**Comment by FelixFaassen-TomTom** [comment]
Hi everyone,

I would be very interested in organizing a separate meeting / workshop to discuss this topic in depth. I think there is a lot of value in having a Map Style Editor close to the renderer and the tile schema specification. 

What would interest me is to learn how the current version of Maputnik is missing features which are available in MapLibre right now (@wipfli already provide a list) and see how those compare to the Style Editor needs of the community members to determine whether there is enough common interest in the community to work on Maputnik based editor under the Maplibre umbrella. It could also help inform the roadmap of Maplibre Renderer and the editor.

Happy to help organize this together if there is interest.

Cheers,

Felix

**Reply by HarelM** [reply]
This sounds like a great idea! Feel free to post it in the slack channel and see who is interested in a kickoff meeting.

**Reply by lseelenbinder** [reply]
Would love to be involved. https://github.com/maputnik/editor/pull/789 is already present as a starting point. 😄

**Reply by vinayakkulkarni** [reply]
Would love to be involved if we could do it from scratch in Vue / Nuxt :)

**Comment by FelixFaassen-TomTom** [comment]
Hi everyone,

Would like to report back. Wednesday the 15th of February Dominika and I had a very nice video chat together with @nyurik on how to move forward on this topic.

As we all know, the crux is that Maputnik is not actively maintained at the moment. We feel that having a workshop and theorizing about work and what-if scenarios on whether we should incubate Maputnik is not helpful now. Incubating an inactive project under the Maplibre umbrella or discussing governance will remain a theoretical exercise when no active work is being done on Maputnik.

So instead of having a workshop, @nyurik recommended to start working on the Maputnik repo by picking up issues which are important to the Maplibre community. He has admin rights on the Maputnik repo, so he can approve PR’s.

Once work on Maputnik is re-activated it also becomes easier to have discussions about governance and project maintenance.

As obvious as this may sound, there are several reasons why Maputnik is not actively maintained up until this point.
One of the reasons, I assume, is that projects (especially inside companies) depending on Maputnik have made an internal fork and make changes to their internal fork instead of the public Maputnik repo. Inside my company this is exactly the case and I think my company is not the only one ;-)

The longer we keep doing this, the harder it becomes to find a rationale for merging changes back to the Maputnik repo. And thus, hindering getting Maputnik into a maintainable state and have it aligned with changes made to the renderer and stylesheet specification.

In my company, we see value in re-activating work on Maputnik and potentially backporting some changes if there is interest with others community members to do the same. That is why I would like to learn from other Maplibre community members that work on projects/products that depend on Maputnik if they see value in contributing their additions to the open version and collaborate on this together.

I’d love to meet you, either in a 1:1 conversation or still organizing a smaller session to learn from what others think. If you are interested, reach out to me here or on the Maplibre slack channel.

I think the longer we wait with working on this, the harder it will become for users to design and visualize their map when using Maplibre and the harder it will become to keep renderer, stylesheet, and editor in sync.

Love to hear what you think and let me know if I have misrepresented things. I am still new to the community, so bear with me ;-)

Cheers,

Felix

**Reply by louwers** [reply]
Sounds like a plan! I sincerely hope Maputnik will spring to live again, it is such a cool project.

**Reply by kevinschaul** [reply]
This sounds like a great plan. Right now a few colleagues of mine are using maputnik as-is. We are using maputnik/desktop, which I've submitted a couple PRs to to improve the installation process.

**Reply by vinayakkulkarni** [reply]
Especially since we moved the style-spec to it's own repo, this change is very welcome! Would love to contribute via code or docs in any ways possible

**Comment by LRathje** [comment]
I just want to mention that MapTiler just launced a new style editor, where you can click and point your way to a new map style. Unfortunately you need to be a paying customer to download the style though...
Link here: [https://cloud.maptiler.com/maps/editor](https://cloud.maptiler.com/maps/editor)

**Reply by HarelM** [reply]
Yup... I received this mail as well, I was hoping MapTiler would push forward something a bit more open source for editing the maplibre style json...
Any comments on this @klokan?

**Comment by sjg-wdw** [comment]
If you had the money, I'd suggest you build out the whole pipeline, the style editor being a piece of that.
It does seem difficult to not have an actively developed style editor.

**Comment by zstadler** [comment]
I've noticed that some of the new MapLibre capabilities are available as plugins, such as [maplibre-contour](https://github.com/onthegomap/maplibre-contour). It is currently impossible to use Maputnik with MapLibre plugins.

Therefore, I would like MapLibre to support its own style editor that:

1. Uses the Maplibre_GL-SJ library rather than Mapbox 
2. Allows the use of MapLibre Plugins

**Reply by nyurik** [reply]
@zstadler what do you mean by "our own" - you mean different from Maputnik, which is already being migrated to MapLibre, and won't be maintained by anyone unless we step in? Or do you propose we invest time in rewriting all of Maputnik's functionality?  I'm all for using plugins of course - the issue is who gets to maintain it :)

**Reply by HarelM** [reply]
I believe the PR to replace Mapbox with MapLibre in maputnik was merged.
I think it's extremely hard to know if and "how much" a plugin is maintained.
We have s list of plugins in the docs here at maplibre which I'll be surprised if half of them are still working with the latest version.
So adding plugings to maputnik seems hard, even harder taking into account the current maintenance effort really happening there...

**Reply by zstadler** [reply]
@HarelM I didn't mean that all plugins would be included by default in the MapLibre editor.
I meant that there would be an simple enough way to select which plugin/s would be loaded into the editor.

**Reply by zstadler** [reply]
@nyurik I meant "our own" in a way that the editor supports the MapLibre community and that the MapLibre organization supports the development and maintenance of the editor.
It would be great if this two-way collaboration would be with Maputnik, which I use and love.
